libbeat/autodiscover/providers/kubernetes/node.go

```diff
@@ -114,7 +115,21 @@ func (n *node) OnUpdate(obj interface{}) {
 // OnDelete ensures processing of node objects that are deleted
 func (n *node) OnDelete(obj interface{}) {
 	n.logger.Debugf("Watcher Node delete: %+v", obj)
-	time.AfterFunc(n.config.CleanupTimeout, func() { n.emit(obj.(*kubernetes.Node), "stop") })
+	node, isNode := obj.(*kubernetes.Node)
+	// We can get DeletedFinalStateUnknown instead of *kubernetes.Node here and we need to handle that correctly. #23385
+	if !isNode {
+		deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
```

Comment:
This is exposing details from the watcher and the client cache in the informers. `DeleteFinalStateUnknown` happens when a watcher misses the delete event, and it sees in a resync that an object is not there anymore. As it doesn't know what was the final state, it includes the last known state, that in our case is probably enough to remove configurations.

Also, handling this in the handler level forces us to repeat some code in each handler we add.

Wdyt about trying to handle it in the watcher directly?

For example [`watcher.enqueue()`](https://github.com/elastic/beats/blob/bf712062819057ba5f958811094a83f8c583e01a/libbeat/common/kubernetes/watcher.go#L196) already does something related to `DeletedFinalStateUnknown` when it calls `cache.DeletionHandlingMetaNamespaceKeyFunc`. We could maybe modify `enqueue` to do something like this:
```
func (w *watcher) enqueue(obj interface{}, state string) {
        // DeletionHandlingMetaNamespaceKeyFunc that we get a key only if the resource's state is not Unknown.
        key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
        if err != nil {
                return
        }
        if deleted, ok := obj.(cache.DeletedFinalStateUnknown); ok {
                w.logger.Debugf(.....)
                obj = obj.Obj
        }
        w.queue.Add(&item{key, obj, state})
}
```